roles/matrix-synapse/tasks/synapse/setup_install.yml

```diff
@@ -47,7 +56,7 @@
       environment:
         DOCKER_BUILDKIT: 1
       when: "matrix_synapse_git_pull_results.changed|bool or matrix_synapse_docker_image_check_result.stdout == ''"
-  when: "matrix_synapse_container_image_self_build|bool"
+  when: "matrix_synapse_container_image_self_build|bool or matrix_synapse_s3_media_store_enabled|bool"
```

Comment:
I think that instead of adding ` or matrix_synapse_s3_media_store_enabled|bool` here, we should instead influence the `matrix_synapse_container_image_self_build` variable.

In `roles/matrix-synapse/defaults/main.yml`:

```diff
-matrix_synapse_container_image_self_build: false
+matrix_synapse_container_image_self_build: "{{ matrix_synapse_s3_media_store_enabled }}"
```

And in `group_vars/matrix_servers`:

```diff
-matrix_synapse_container_image_self_build: "{{ matrix_architecture not in ['arm64', 'amd64'] }}"
+matrix_synapse_container_image_self_build: "{{ matrix_architecture not in ['arm64', 'amd64'] or matrix_synapse_s3_media_store_enabled }}"
```

------

For a while, I thought that redefining `matrix_synapse_container_image_self_build` is dirtier, but looking a few lines below (for the "Ensure Synapse Docker image is pulled"), it seems like your change introduces a bug.

The "Ensure Synapse Docker image is pulled" task's `when` either needs to be changed:

```diff
-when: "not matrix_synapse_container_image_self_build"
+when: "not matrix_synapse_container_image_self_build or not matrix_synapse_s3_media_store_enabled"
```

.. or, we need to influence the `matrix_synapse_container_image_self_build` variable as described above and then it will all work.

Reply:
I like the idea to influence the `self_build` var. I already added a disclaimer in the docs to not disabling or overwriting the Synapse `self-build` var.